### PR TITLE
[WIP] Use micromamba for all platforms

### PR DESCRIPTION
**Pull request by isuruf**
Checklist
* [ ] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by wolfv** [comment]
Hi @isuruf I just noticed this PR. Do you think the failures are related to a bug in micromamba? Would be happy to look into it & fixing it.

**Comment by isuruf** [comment]
> Do you think the failures are related to a bug in micromamba? 

Yes, but it's unclear why.

**Comment by wolfv** [comment]
Maybe tempfolder doesn't exist and mamba doesn't create it. I'll have a look on Monday.

We recently changed slightly how noarch packages are compiled (it creates a temporary wrapper script to call the pyc compilation now).

**Comment by isuruf** [comment]
> Maybe tempfolder doesn't exist and mamba doesn't create it. I'll have a look on Monday.

Yes, that's one reason. constructor sets `TMP` env variable to a nonexistent folder, but micromamba doesn't create it.

If I remove that temporarily, there's a bunch of warnings like,
```
/tmp/mambafZMK04KHgU2: line 4: /root/miniforge3//.libgomp-post-link.sh: No such file or directory
```
micromamba should first check to see if these files exist before trying to call them.

**Comment by wolfv** [comment]
I'll have a deeper look on monday. Normally we do check for script existence here:

https://github.com/mamba-org/mamba/blob/d27e65b5f5a1dae7281141bc8200bd4528d3d7d8/src/link.cpp#L536-L541

Not sure why this tries to execute non-existent scripts.

I was thinking that whoever sets `TMP` should also create it, but I can change that behavior. I recently checked the Python source code and as far as I understand they don't create TMP either (they fall back to `/tmp` instead here https://github.com/python/cpython/blob/3cf0833f42ebde24f6435b838785ca4f946b988f/Lib/tempfile.py#L187)

**Comment by isuruf** [comment]
Might be a bad interaction with QEMU.

**Comment by isuruf** [comment]
@wolfv, thoughts?

**Comment by wolfv** [comment]
I just looked at it again, and it's weird that the path is not `$PREFIX/bin/.thatfile` as that's how it should be computed here:

https://github.com/mamba-org/mamba/blob/243bf801e99fc2215f88fd87c66574aeb85a52e9/src/link.cpp#L530-L534

and get_bin_short_path should always just return `"bin"` 

https://github.com/mamba-org/mamba/blob/3b71f37525fcf1825dfe8aea74a9b3dab0dff915/src/transaction_context.cpp#L43-L50

That might be a first hint at where the bug is...
I have tried to build the installer locally on a Fedora machine but somehow this docker doesn't start. 

Edit just figured out I can download the artifact from Github. Will run this on my pinephone :)

**Comment by wolfv** [comment]
I can reproduce it on the pinephone so it doesn't seem to be related to QEMU. Will check some more tomorrow (today is a holiday in berlin).

**Comment by wolfv** [comment]
I checked with the upcoming micromamba release and it didn't seem to have the same problem. So let's wait a tiny bit more (until the windows build of micromamba 0.8.0 is fixed) and try again.

**Comment by isuruf** [comment]
Fails with,
```
ERROR   ssl_verify is enabled but no ca certificates found
Warning: setting fallback $MAMBA_ROOT_PREFIX to "/root/micromamba".

You have not set a $MAMBA_ROOT_PREFIX environment variable.
To permanently modify the root prefix location, either set the MAMBA_ROOT_PREFIX environment variable, or use
  micromamba shell init ... 
to initialize your shell, then restart or source the contents of the shell init script.
```

**Comment by isuruf** [comment]
Looks like new micromamba looks for ssl certificates even if `--offline` is set in the install command.

**Comment by wolfv** [comment]
sorry. Will fix that now, and push out a new release soon

**Comment by isuruf** [comment]
Thanks

**Comment by wolfv** [comment]
yay, looks like `0.8.2` fixes the certificate problems.

**Comment by wolfv** [comment]
argh. will have to run additional tests.

File: scripts/build.sh [hunk 1]
```diff
@@ -33,11 +33,10 @@ cp LICENSE $TEMP_DIR/
...
 ls -al $TEMP_DIR
 
-if [[ $(uname -r) != "$ARCH" ]]; then
-    if [[ "$ARCH" == "arm64" ]]; then
-        CONDA_SUBDIR=osx-arm64 conda create -n micromamba micromamba=0.6.5 -c https://conda-web.anaconda.org/conda-forge --yes
-        EXTRA_CONSTRUCTOR_ARGS="$EXTRA_CONSTRUCTOR_ARGS --conda-exe $CONDA_PREFIX/envs/micromamba/bin/micromamba --platform osx-arm64"
-    fi
+if [[ $TARGET_PLATFORM != win-* ]]; then
+  CONDA_SUBDIR=$TARGET_PLATFORM conda create -n micromamba micromamba=0.8.2 -c conda-forge --yes
+  MICROMAMBA_FILE=$(find $CONDA_PREFIX/pkgs -iname micromamba)
```
Comment (isuruf) [inline scripts/build.sh:38]: Why use it from `pkgs`?

Comment (wolfv) [reply scripts/build.sh:38]: because when I tried on the server, the unmodified micromamba worked fine. There is a prefix replacement going on, and I think it might break micromamba for some reason.

Comment (isuruf) [reply scripts/build.sh:38]: Okay. Thanks. Shall we remove the hash from micromamba so that we have stable download links like `https://anaconda.org/conda-forge/micromamba/$VERSION/download/$TARGET_PLATFORM/micromamba-$VERSION-0.tar.bz2`

Comment (wolfv) [reply scripts/build.sh:38]: it's a good idea, yeah!

**Comment by wolfv** [comment]
btw @isuruf not sure if you're aware, but I've been also working on the boa-forge stuff (should probably rename it to micromamba-forge) that has a github actions based pipeline to build micromamba and all dependencies (since today it also supports linux-aarch64 and osx-arm64 using cross-compilation).
The binaries (for linux) are a little bit smaller because we can configure curl and libarchive with fewer features (I think boa-forge has ~8Mb and conda-forge 12Mb). Anyways, I need to look into how we can make these binaries smaller (maybe some different compiler flags could help, or turning on MinSizeRel in cmake etc).

**Comment by wolfv** [comment]
Link: https://github.com/mamba-org/boa-forge/releases

**Comment by wolfv** [comment]
PS I don't think we should use the outputs from boa-forge here. It's nice to dog-food the conda-forge produced binaries and gives more control to the community!

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -149,30 +169,31 @@ jobs:
...
-          EXT=exe
         fi
         # Copy for latest release
         cp build/$MINIFORGE_NAME-*-$OS_NAME-$ARCH.$EXT build/$MINIFORGE_NAME-$OS_NAME-$ARCH.$EXT
         ls -alh build
       shell: bash
 
     - name: Upload miniforge to Github artifact
-      if: ${{ !startsWith(github.ref, 'refs/tags/') }}
+      if: ${{ failure() }}
```
Comment (isuruf) [inline .github/workflows/ci.yml:196]:
```suggestion
      if: ${{ !startsWith(github.ref, 'refs/tags/') }}
```

**Comment by wolfv** [comment]
Hi @isuruf, I finally got this to work again with micromamba ... there were some additional checks that were introduced recently that broke with no proper `repodata_record.json` available. Now we properly read the urls from the `urls` file that comes with conda constructor. I did have to change the order in which `conda constructor` extracts data though (after my changes it first extracts $PRECONDA and then the tar-balls): https://github.com/conda/constructor/pull/450

Running tests here: https://github.com/wolfv/miniforge/runs/2570759561?check_suite_focus=true

Any idea what's happening with conda == 4.8.5 on ppc64le?

**Comment by hmaarrfk** [comment]
Can you summarize, in lay terms, what this brings to miniforge? A smaller installer? Faster build times? Smaller build matrix?

**Comment by wolfv** [comment]
@hmaarrfk on osx-arm64 I think we can only use micromamba currently, as there is no conda-standalone available due to limitations with cross-building (as far as I understand). Using micromamba everywhere might give a bit more homogeneity.

Knowing that micromamba works on all platforms could also lead to smaller constructors for other projects (e.g. R, Ruby, Rust ...) as they don't need to ship Python + conda-standalone with the constructor.

But I think that for miniforge itself there is not a super-huge benefit (actually an increase of ~3Mb for some of the installers).

Anyways, for me personally it's nice to see that this works fine now :)

Maybe @isuruf has some additional reasons.

Btw. @isuruf feel free to remove / squash my commits. Sorry for cluttering the history so much here -- I am completely fine with you taking all the credit.

**Comment by isuruf** [comment]
> Knowing that micromamba works on all platforms could also lead to smaller constructors for other projects (e.g. R, Ruby, Rust ...) as they don't need to ship Python + conda-standalone with the constructor.

This is not true. `micromamba` and `conda-standalone` both are roughly the same size so there's no benefit in size for other projects.

Main reason for me is to avoid `conda-standalone` which has a ton of dependencies and we are not mentioning the licenses to any of the dependencies. With `micromamba` it's easier to figure out the licensing.

@hmaarrfk, let me know what you think and I'll merge this after cleaning it up a bit.

**Comment by hmaarrfk** [comment]
I think it is cool.

Honestly, I'm soon going to propose including mamba as default just because the speed is so advantageous.

**Review by hmaarrfk — APPROVED** [review]

File: scripts/build.sh [hunk 3]
```diff
@@ -3,21 +3,26 @@
...
+
 if [[ "$(uname)" == "Darwin" ]]; then
-    conda install -y coreutils -c conda-forge -c defaults --override-channels
+    conda install -y coreutils -c conda-forge --override-channels
 elif [[ "$(uname)" == MINGW* ]]; then
-    conda install -y "nsis=3.01" -c conda-forge -c defaults --override-channels
+    conda install -y "nsis=3.01" -c conda-forge --override-channels
 fi
-pip install git+git://github.com/conda/constructor@3.2.1#egg=constructor --force --no-deps
+pip install git+git://github.com/wolfv/constructor@69eaa20b12ea1020438f645ec834f47ac9d3c3a7#egg=constructor --force --no-deps
```
Comment (jakirkham) [inline scripts/build.sh:25]: Asked upstream about getting this change reviewed ( https://github.com/conda/constructor/pull/450 )
